### ISSUE TITLE
Test wanted: ecRecover failure for py-evm

### ISSUE THREAD

**Issue by holiman** [ethereum/tests#941]
Reported by @guidovranken.  cc @carver @pipermerriam 

Here's a repro of a testcase that apparently fails on py-evm, but succeeds on geth and nethermind. The test calls `ecRecover`, and the signature recovery fails on geth and nethermind, but succeeds on py-evm
```json
{
 "ecRecoverTest": {
  "env": {
   "currentCoinbase": "b94f5374fce5edbc8e2a8697c15331677e6ebf0b",
   "currentDifficulty": "0x20000",
   "currentGasLimit": "0x26e1f476fe1e22",
   "currentNumber": "0x1",
   "currentTimestamp": "0x3e8",
   "previousHash": "0x0000000000000000000000000000000000000000000000000000000000000000",
   "currentBaseFee": "0x10"
  },
  "pre": {
   "0x000000000000000000000000000000000000ff0a": {
    "code": "0x7f6b8d2c81b11b2d699528dde488dbdf2f94293d0d33c32e347f255fa4a6c1f0a9600052601b6020527f79be667ef9dcbbac55a06295ce870b07029bfcdb2dce28d959f2815b16f817986040527f6b8d2c81b11b2d699528dde488dbdf2f94293d0d33c32e347f255fa4a6c1f0a9606052602060006080600060006001620f4240f150600051600055",
    "storage": {},
    "balance": "0xffffffff",
    "nonce": "0x0"
   },
   "0xa94f5374fce5edbc8e2a8697c15331677e6ebf0b": {
    "code": "0x",
    "storage": {},
    "balance": "0xde0b6b3a7640000",
    "nonce": "0x0"
   }
  },
  "transaction": {
   "gasPrice": "0x10",
   "nonce": "0x0",
   "to": "0x000000000000000000000000000000000000ff0a",
   "data": [
    ""
   ],
   "gasLimit": [
    "0x989680"
   ],
   "value": [
    "0x0"
   ],
   "secretKey": "0x45a915e4d060149eb4365960e6a7a45f334393093061116b197e3240065ff2d8"
  },
  "out": "0x",
  "post": {
   "London": [
    {
     "hash": "b520c02c1dedacb96574e47fe27a3053e9d0f2b9d52dadd3702a00466532ac31",
     "logs": "1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347",
     "indexes": {
      "data": 0,
      "gas": 0,
      "value": 0
     }
    }
   ]
  }
 }
}
```

The code is 
```
	copy(v, hexutil.MustDecode("0x000000000000000000000000000000000000000000000000000000000000001b"))
	copy(r, hexutil.MustDecode("0x79be667ef9dcbbac55a06295ce870b07029bfcdb2dce28d959f2815b16f81798"))
	copy(s, hexutil.MustDecode("0x6b8d2c81b11b2d699528dde488dbdf2f94293d0d33c32e347f255fa4a6c1f0a9"))
	copy(hash, hexutil.MustDecode("0x6b8d2c81b11b2d699528dde488dbdf2f94293d0d33c32e347f255fa4a6c1f0a9"))

	a.Mstore(hash, 0)
	a.Mstore(v, 32)
	a.Mstore(r, 64)
	a.Mstore(s, 96)

	a.Call(big.NewInt(1_000_000),
		1,
		big.NewInt(0),   // value
		big.NewInt(0),   // inoffset
		big.NewInt(128), // insize
		big.NewInt(128),
		big.NewInt(128), // outsize
	)
	a.Op(ops.RETURNDATASIZE)
	a.Op(ops.SSTORE)
``` 
so it just creates the input `6b8d2c81b11b2d699528dde488dbdf2f94293d0d33c32e347f255fa4a6c1f0a9000000000000000000000000000000000000000000000000000000000000001b79be667ef9dcbbac55a06295ce870b07029bfcdb2dce28d959f2815b16f817986b8d2c81b11b2d699528dde488dbdf2f94293d0d33c32e347f255fa4a6c1f0a9` and invokes ecrecover, then sets a slot depending on the returndatasize (should be `0`) and the return value from the call  (should be `1`).

**Comment by holiman** [comment]
@carver can you ACK whether py-evm indeed fails this test or not?

**Comment by MariusVanDerWijden** [comment]
Succeeds on OpenEthereum and Besu too

**Comment by guidovranken** [comment]
> Succeeds

The test, or the ECRECOVER call?

**Comment by MariusVanDerWijden** [comment]
The test, it executes in the same way that geth and nethermind execute it

**Comment by carver** [comment]
> @carver can you ACK whether py-evm indeed fails this test or not?

Yes, I believe @kclowes is working on this and it has been reproduced locally. She or I will link the PR when it's posted to github.

**Comment by qbzzt** [comment]
Did you mean for `hash` and `s` to be the same value? If so, here's a filler file that implements that test.

[ecRecoverFiller.yml.txt](https://github.com/ethereum/tests/files/7125309/ecRecoverFiller.yml.txt)

**Comment by kclowes** [comment]
Here's the PR for py-evm: https://github.com/ethereum/py-evm/pull/2024. I'll make sure these test cases are pulled in too!

**Comment by holiman** [comment]
That fix doesn't look like it fixes the right thing. The data size is correct, isn't it?

**Comment by kclowes** [comment]
Based on the email that came through, I thought it shouldn't succeed because the r and s values were too long. Is that wrong? I'm new to the EVM side of things :)

**Comment by holiman** [comment]
Yes the actual values provided in the email didn't match up with what happened when executing the code that was also provio. Could you try the values given in this issue description?

**Comment by kclowes** [comment]
Ah, okay. Yep, will do!

**Comment by holiman** [comment]
That is, how does py-evm handle the input `6b8d2c81b11b2d699528dde488dbdf2f94293d0d33c32e347f255fa4a6c1f0a9000000000000000000000000000000000000000000000000000000000000001b79be667ef9dcbbac55a06295ce870b07029bfcdb2dce28d959f2815b16f817986b8d2c81b11b2d699528dde488dbdf2f94293d0d33c32e347f255fa4a6c1f0a9` to ecrecover?

**Comment by kclowes** [comment]
@holiman py-evm does send back an address (3f17f1962b36e491b30a40b2405849e597ba5fb5) with the input from above. It isn't clear to me though why those values/that input should fail?

**Comment by kclowes** [comment]
~~Actually, looks like r and s are too big?~~


EDIT: I deleted the content of this comment because it was wrong. I copy/pasta'ed the wrong secp256k1n locally 🤦‍♀️

**Comment by kclowes** [comment]
@holiman why should that input fail?

**Comment by holiman** [comment]
I was hoping you'd tell me :)I'l try to figure it out.

**Comment by kclowes** [comment]
Ah, okay. I’ll keep looking too. I didn’t want to spend a bunch of time if you knew!

**Comment by holiman** [comment]
EDIT: nevermind, investigating...

**Comment by holiman** [comment]
I don't see why that isn't rejected by py-evm at this point: https://github.com/kclowes/py-evm/blob/master/eth/precompiles/ecrecover.py#L47

**Comment by holiman** [comment]
As far as I can tell, it has something to do with points at infinity. It fails this check https://github.com/ethereum/go-ethereum/blob/master/crypto/secp256k1/libsecp256k1/src/modules/recovery/main_impl.h#L120

I updated the statetest, please try it on py-evm.

**Comment by winsvega** [comment]
Does those tests cover the case enough?

**Comment by holiman** [comment]
cc @kclowes does the test https://github.com/ethereum/tests/pull/973 'work' on pyevm? 

Also, is there a statetest which implements the one I posted in the issue description @winsvega ?

**Comment by winsvega** [comment]
Yes there is

`GeneralStateTests/stTransactionTest/PointAtInfinityECRecover.json`

That calls ECRecover precompile

**Comment by kclowes** [comment]
@holiman I _think_ so, but need to double check that I'm running the right thing. I should be able to get to it tomorrow.

**Comment by winsvega** [comment]
status?

**Comment by kclowes** [comment]
Fix has been merged and released in py-evm (v0.5.0a2) via https://github.com/ethereum/eth-keys/pull/76. I am working on getting the changes into eth-account now! This issue can probably be closed though. Thanks!
